### PR TITLE
PXB-8.0-2740 PXB displays the encryption key when the backup is decrypted…

### PR DESCRIPTION
**Pull request by rahulmalik87**
… and decompressed

https://jira.percona.com/browse/PXB-2740

Problem:
PXB displays the encryption key when the backup is decrypted and decompressed

Analysis:
PXB uses system function to call xbcrypt with encryption password to decrypt
the password. system function creates sub process. password is only
removed by sub process . The main process shows the encrypted password
in ps

Fix:
set the enviorment variable XBCRYPT_ENCRYPTION_PASSWORD in pxb and use
that variable inside xbcrypt

### CODE REVIEW

File: storage/innobase/xtrabackup/src/xbcrypt_common.cc [hunk 1]
```diff
@@ -305,3 +305,10 @@ gcry_error_t xb_crypt_encrypt(gcry_cipher_hd_t cipher_handle, const uchar *from,
 
   return 0;
 }
+
+void get_env_value(char *&var, const char *env) {
+  char *val;
+  if (var == nullptr && (val = getenv(env)) != nullptr) {
+    var = my_strdup(PSI_NOT_INSTRUMENTED, val, MYF(MY_FAE));
```
Comment (satya-bodapati) [inline storage/innobase/xtrabackup/src/xbcrypt_common.cc:312]: where do we free this?

File: storage/innobase/xtrabackup/src/backup_copy.cc [hunk 2]
```diff
@@ -2584,9 +2585,7 @@ bool decrypt_decompress_file(const char *filepath, uint thread_n) {
   if (ends_with(filepath, ".xbcrypt") && opt_decrypt) {
     cmd << " | xbcrypt --decrypt --encrypt-algo="
         << xtrabackup_encrypt_algo_names[opt_decrypt_algo];
-    if (xtrabackup_encrypt_key) {
-      cmd << " --encrypt-key=" << xtrabackup_encrypt_key;
-    } else {
+    if (xtrabackup_encrypt_key == nullptr) {
```
Comment (satya-bodapati) [inline storage/innobase/xtrabackup/src/backup_copy.cc:2588]: can we do setenv() here?
if (xtrabckup_encrypt_key) {
setenv();
} else {
--file
}

Comment (satya-bodapati) [inline storage/innobase/xtrabackup/src/backup_copy.cc:2588]: also, I think it is better to unset it after usage.
get old value (getnv()), save it. and restore it after system() call.

**Review by satya-bodapati — CHANGES_REQUESTED** [review]
also better to do GCA and make PR for 2.4

File: storage/innobase/xtrabackup/src/backup_copy.cc (hunk 2 above)
Comment (satya-bodapati) [reply storage/innobase/xtrabackup/src/backup_copy.cc:2588]: on second thought, I think, the ENV change done by this xtrabackup process will not be visible to other xtrabackup processes. So pxb exit will auto unset it.

Please verify

Comment (rahulmalik87) [reply storage/innobase/xtrabackup/src/backup_copy.cc:2588]: @satya-bodapati , if we do setenv inside decrypt_decompress_file we have to do it for each file. better to do it outside of this function so global environment will be called once .

Comment (rahulmalik87) [reply storage/innobase/xtrabackup/src/backup_copy.cc:2588]: correct, pxb exists with auto unset.

File: storage/innobase/xtrabackup/src/backup_copy.cc [hunk 3]
```diff
@@ -2655,6 +2654,10 @@ static void decrypt_decompress_thread_func(datadir_thread_ctxt_t *ctxt) {
       continue;
     }
 
+    if (xtrabackup_encrypt_key) {
+      setenv("XBCRYPT_ENCRYPTION_KEY", xtrabackup_encrypt_key, 1);
+    }
+
```
Comment (satya-bodapati) [inline storage/innobase/xtrabackup/src/backup_copy.cc]: here too, we seem to set after every file, this should be set before the while loop isn't it?

**Comment by rahulmalik87** [comment]
@satya-bodapati  test results.



==============================================================================
TEST                                   WORKER    RESULT     TIME(s) or COMMENT
------------------------------------------------------------------------------
check_for_passwords                      w1	[passed]    60
==============================================================================
Spent 60 of 86 seconds executing testcases

SUMMARY: 1 run, 1 successful, 0 skipped, 0 failed

**Review by satya-bodapati — APPROVED** [review]

**Pull request merged**